executor/write.go

```diff
@@ -882,23 +887,52 @@ func getRecordIDs(ctx sessionctx.Context, t table.Table, rows [][]types.Datum) (
 			}
 		}
 		for _, row := range rows {
-			recordIDs = append(recordIDs, row[handleCol.Offset].GetInt64())
+			handles = append(handles, row[handleCol.Offset].GetInt64())
 		}
 	} else {
 		for range rows {
-			recordID, err := t.AllocAutoID(ctx)
+			handle, err := t.AllocAutoID(ctx)
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
-			recordIDs = append(recordIDs, recordID)
+			handles = append(handles, handle)
 		}
 	}
-	return recordIDs, nil
+	return handles, nil
+}
+
+func getOldValues(ctx sessionctx.Context, t table.Table, handles []int64) (map[string][]byte, error) {
+	batchKeys := make([]kv.Key, 0, len(handles))
+	for _, handle := range handles {
+		batchKeys = append(batchKeys, t.RecordKey(handle))
+	}
+	values, err := ctx.Txn().GetSnapshot().BatchGet(batchKeys)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return values, nil
+}
+
+func encodeNewRow(ctx sessionctx.Context, t table.Table, row []types.Datum) ([]byte, error) {
+	colIDs := make([]int64, 0, len(row))
+	skimmedRow := make([]types.Datum, 0, len(row))
+	for _, col := range t.WritableCols() {
+		if !tables.CanSkip(t.Meta(), col, row[col.Offset]) {
+			colIDs = append(colIDs, col.ID)
+			skimmedRow = append(skimmedRow, row[col.Offset])
+		}
+	}
+	newRowValue, err := tablecodec.EncodeRow(ctx.GetSessionVars().StmtCtx, skimmedRow, colIDs, nil, nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return newRowValue, nil
 }
 
 // getKeysNeedCheck gets keys converted from to-be-insert rows to record keys and unique index keys,
 // which need to be checked whether they are duplicate keys.
-func getKeysNeedCheck(ctx sessionctx.Context, t table.Table, rows [][]types.Datum) ([][]keyWithDupError, error) {
+func getKeysNeedCheck(ctx sessionctx.Context, t table.Table, rows [][]types.Datum) ([][]keyWithDupError,
```

Comment:
The on dup update values may be different than the new row value, needs to be checked too.